types/recoil/lib/core/recoilValue.d.ts

```diff
@@ -0,0 +1,21 @@
+import { NodeKey } from './state';
+
+export class AbstractRecoilValue<T> {
+    tag: 'Writeable';
+    key: NodeKey;
+    constructor(newKey: NodeKey);
+}
+
+export class AbstractRecoilValueReadonly<T> {
+    tag: 'Readonly';
+    key: NodeKey;
+    constructor(newKey: NodeKey);
+}
+
+export class RecoilState<T> extends AbstractRecoilValue<T> {}
```

Comment:
`RecoilState` should be both covariance and contravariance. because it is both readable and writable.

# Example Problem
there is no protection to assigning `RecoilState<number>` to `RecoilState<string>`

```ts
let stringAtom = atom<string>({ key: 's', default: 'some string' })
let numberAtom = atom<number>({ key: 'n', default: 45 })

stringAtom = numberAtom // no error. it should detect error

const s: string = useRecoilValue(stringAtom)
// at runtime s is number
```

# Solution
to fix this problem you can use a guard hack pattern. [example](https://github.com/forminator/react-wire/blob/fa6e5631610e6a6424dbe33e81b34e9cfe59e116/index.d.ts#L67)

types (`StateWire` is very similar to `RecoilState`):
```ts
declare interface StateWireGuard<V> {
  ' state-wire': StrictGuard<V>;
}
export declare interface StateWire<V> extends StateWireGuard<V> {
  getValue(): V;
  setValue(value: Defined<V>): void;
  subscribe(callback: (value: Defined<V>) => void): () => void;
}
declare type StrictGuard<T> = [T, CB<T>];
declare type Defined<T> = T extends undefined ? never : T;
declare type CB<T> = (t: T) => void;

```
sample usage:
```
let stringWire: StateWire<string> = {} as any
let numberWire: StateWire<number> = {} as any
let numberOrStringWire: StateWire<number|string> = {} as any

numberWire = stringWire // error, ok
numberWire = numberOrStringWire // error, ok, covariance check
numberOrStringWire = numberWire // error, ok, contravariance check

```
errors:
```
Type 'StateWire<string>' is not assignable to type 'StateWire<number>'.
  The types returned by 'getValue()' are incompatible between these types.
    Type 'string' is not assignable to type 'number'.
Type 'StateWire<string | number>' is not assignable to type 'StateWire<number>'.
  The types returned by 'getValue()' are incompatible between these types.
    Type 'string | number' is not assignable to type 'number'.
      Type 'string' is not assignable to type 'number'
Type 'StateWire<number>' is not assignable to type 'StateWire<string | number>'.
  Types of property '' state-wire'' are incompatible.
    Type 'StrictGuard<number>' is not assignable to type 'StrictGuard<string | number>'.
      Type 'CB<number>' is not assignable to type 'CB<string | number>'.
        Type 'string | number' is not assignable to type 'number'.
          Type 'string' is not assignable to type 'number'.
```

Reply:
Good catch! The issue is that we're not using `T` anywhere in the definition of `AbstractRecoilValue<T>` and `AbstractRecoilValueReadonly<T>` so TS treats them as structurally equal. To fix it I'm going to add a `valTag: T` property

Reply:
the issue with `RecoilValueReadOnly` fixed. but `RecoilState` has a problem yet.

## example problem
```ts
const stringAtom: RecoilState<string> = {} as any;
const stringOrNumberAtom: RecoilState<string|number> = stringAtom // no error. it should detect error, it should check contravariance. because RecoilState is writable

const [value, setValue] = useRecoilState<string|number>(stringOrNumberAtom)
setValue(5) // no error, ok, SetterOrUpdater<string|number> should accept number

const value2 = useRecoilValue(stringAtom) // expect value2 should be string, but it is 5
```
## solution
How to fix this? add contravariance prop:

```ts
declare type CB<T> = (t: T) => void;
export class AbstractRecoilValue<T> {
    tag: 'Writeable';
    valTag: T;
    setTag: CB<T>;
    key: NodeKey;
    constructor(newKey: NodeKey);
}
```

## advice

if `valTag` doesn't exist in js add space at the start of the key. and also change prop name to some unintuitive name

```ts
declare type CB<T> = (t: T) => void;
export class AbstractRecoilValue<T> {
    tag: 'Writeable';
    ' covariance-value': T; // mind space at the start
    ' contravariance-value': CB<T>; // mind space at the start
    key: NodeKey;
    constructor(newKey: NodeKey);
}

export class AbstractRecoilValueReadonly<T> {
    tag: 'Readonly';
    ' covariance-value': T; // mind space at the start
    key: NodeKey;
    constructor(newKey: NodeKey);
}

```

Reply:
by replacing `T` with `[T]` we can protect props from removing by `never` type. without this `[T]`, the `RecoilValueReadOnly<never>` is equal to `RecoilValueReadOnly<any>`

```ts
declare type CB<T> = (t: T) => void;
export class AbstractRecoilValue<T> {
    tag: 'Writeable';
    ' covariance-value': [T]; // never type protectoin
    ' contravariance-value': [CB<T>]; // never type protectoin
    key: NodeKey;
    constructor(newKey: NodeKey);
}

export class AbstractRecoilValueReadonly<T> {
    tag: 'Readonly';
    ' covariance-value': [T]; // never type protectoin
    key: NodeKey;
    constructor(newKey: NodeKey);
}
```

Reply:
I think with these new types, we can remove `tag` props and use `RecoilValueReadOnly` in read-only functions like `useRecoilValue`

see [playground]

final types:
```ts
export class AbstractRecoilValue<T> {
    ' covariance-value': [T]; // never type protectoin
    ' contravariance-value': [CB<T>]; // never type protectoin
    key: NodeKey;
    constructor(newKey: NodeKey);
}

export class AbstractRecoilValueReadonly<T> {
    ' covariance-value': [T]; // never type protectoin
    key: NodeKey;
    constructor(newKey: NodeKey);
}
```

[playground]: https://www.typescriptlang.org/play/#code/C4TwDgpgBAcg9gEwgaQiKBeKBnYAnASwDsBzAWACgkBjAGwEM9pRIoBhAIQB4AVAPkxQAFMABcUHgEpMAgG5wCCANyUIADzBw8wKHXrZsUAIIAjXHnrVgAJQjUFtAGr1aAVwi8BAb0pQ-UAHJdOFlGAnoiaggAWlC3CADxAG0eAF0lKAB6TKgiCFkIPCgWaDA8OGA7YAUiX38g+yJ8elDCCKjYl3dEqCTOT3SsnLyCopKoMoqqmrq-AGs0cXgkVBAVCn9gonNXKy0hPIB3VaXEFDRJdYBfSlUNLR09A2MzZqtbewInLohbegQ4ERaCBPFAfBt6sFWuFIjE4t1kmkMtlcvlCsVwKVypU9sRZlAFiBTis0OtNo0dns8AcIMdFrAzqtLpQbhQ7pptLoGM8Pg5nPE-ggAPJAkH8KDqSpEBCGUzmSw2Ox8n6CwHA0FeVnsh4Y1i8r789ygrD674CiD-EXq8UAHygpoAysB6JVPOt1BzHtzDI7na7xZKINLZa8LO8lQafhqtWyKBSdOUjNUALbiU2G34W4WirjmYgkARYTVQfQlohrSjxqCHehJuCp+0R2hOl0ePOkQtgq4lwwRCtxwG4XLYOsN32trhEVzJkyFG3tguCYulvvI4bT2dFLQ4fD524UD068YOiDASp4IV4ACqYAQrbwxuEcUvN7v5-EQiE1FceDw-PEUgyBI0h2oBGByAoyj7oenIAGauJEwAEICUCuNgmafGaRr8EITCYRmaZNhmniSAB7r3HBCFWMhRCoeh47+nwuFNi2lSEZhrEePwpG9DwAA0UAnmehQvre94DOs+5oRhyrxLhcCjpIlDSemPxCDWinKfRLF+hA8mKUMEq-loAlwHMAkqTpra6BEAQ6JYURgAmWZQGq6AuvWWkyV8nHqbWKZKbGlkcbpk4boUTEaQFhmFOUeCmXMXkMR4U4zuidoLpF-n1tIKKxSZrnmVCYTtNAEQIFszTQqVugABZ2AsCBJURUapZuWUGXlxnxYVzX4a14VFBlu4dn5nXDHARlxQlAlEAQUSmaKxVtLCdUNRATVBdp-XxGFaV4Ex2zjVNBVmctMJRGt1CNZQQA